Fix multiple issues for Hacktoberfest 2025 - Year in Progress

TechnoBlogger14o3 opened the pull request:
Fixes:
- Issue #24: Refactor index.html by Separating CSS and JavaScript into External Files
- Issue #22: Enhancement: Adding icons of moon and sun instead of writing 'switching to light mode'
- Issue #11: Improve Dark Mode Toggle UI/UX
- Issue #39: Add GitHub Logo to Repository Link
- Issue #46: CSS and JavaScript Enhancements for Optimized Performance and Maintainability

Changes:
- Refactored HTML by separating JavaScript into external files (theme.js, animation.js)
- Enhanced theme toggle with moon/sun icons and improved UI/UX
- Added GitHub logo with enhanced styling and hover effects
- Improved overall UI with modern gradients, glassmorphism effects, and animations
- Optimized performance with will-change properties and reduced motion support
- Enhanced accessibility with proper ARIA labels and tooltips

UI Improvements:
- Modern gradient theme toggle button with hover animations
- Enhanced GitHub link with glassmorphism styling
- Improved form elements with better focus states and transitions
- Enhanced progress image with hover effects and rounded corners
- Modernized footer with gradient background
- Better visual hierarchy with consistent spacing and shadows

Code Organization:
- Separated JavaScript into logical modules (theme.js, animation.js)
- Created enhanced-styles.css for better maintainability
- Improved code structure and readability
- Added performance optimizations and accessibility features

New Features:
- Enhanced theme toggle with smooth icon transitions
- Improved GitHub repository link with better visual feedback
- Modern glassmorphism design elements throughout
- Better responsive design with improved mobile experience
- Enhanced animations and micro-interactions

All changes maintain existing functionality while significantly improving user experience, code maintainability, and visual appeal.

TechnoBlogger14o3 committed:
Fix multiple issues for Hacktoberfest 2025 - Year in Progress

Fixes:
- Issue #24: Refactor index.html by Separating CSS and JavaScript into External Files
- Issue #22: Enhancement: Adding icons of moon and sun instead of writing 'switching to light mode'
- Issue #11: Improve Dark Mode Toggle UI/UX
- Issue #39: Add GitHub Logo to Repository Link
- Issue #46: CSS and JavaScript Enhancements for Optimized Performance and Maintainability

Changes:
- Refactored HTML by separating JavaScript into external files (theme.js, animation.js)
- Enhanced theme toggle with moon/sun icons and improved UI/UX
- Added GitHub logo with enhanced styling and hover effects
- Improved overall UI with modern gradients, glassmorphism effects, and animations
- Optimized performance with will-change properties and reduced motion support
- Enhanced accessibility with proper ARIA labels and tooltips

UI Improvements:
- Modern gradient theme toggle button with hover animations
- Enhanced GitHub link with glassmorphism styling
- Improved form elements with better focus states and transitions
- Enhanced progress image with hover effects and rounded corners
- Modernized footer with gradient background
- Better visual hierarchy with consistent spacing and shadows

Code Organization:
- Separated JavaScript into logical modules (theme.js, animation.js)
- Created enhanced-styles.css for better maintainability
- Improved code structure and readability
- Added performance optimizations and accessibility features

New Features:
- Enhanced theme toggle with smooth icon transitions
- Improved GitHub repository link with better visual feedback
- Modern glassmorphism design elements throughout
- Better responsive design with improved mobile experience
- Enhanced animations and micro-interactions

All changes maintain existing functionality while significantly improving
user experience, code maintainability, and visual appeal.

DhanushNehru commented on public/js/theme.js:
```diff
@@ -0,0 +1,80 @@
+
+themeToggleButton.addEventListener("click", () => updateTheme(false));
+
+setInterval(updateProgressImg, 500);
```
> Updating the progress image every 500ms is unnecessarily frequent and could impact performance. Consider reducing the frequency to every 5-10 seconds or using a more efficient update mechanism.

Hi @TechnoBlogger14o3 , apart from this fix other suggestions from copilot

TechnoBlogger14o3 replied:
Hi @DhanushNehru 
🔧 Issues Fixed:
Performance Issue ✅
Problem: Updating progress image every 500ms was unnecessarily frequent
Solution: Reduced frequency to 5000ms (5 seconds) for better performance
Impact: Significantly reduces API calls and improves overall performance
Security Issue ✅
Problem: Using innerHTML for icon injection created potential XSS vulnerabilities
Solution: Replaced with safer DOM manipulation using querySelector and className
Impact: Eliminates security risks while maintaining functionality
Code Duplication ✅
Problem: HTML injection logic was duplicated in multiple places
Solution: Created updateThemeIcon() helper function to centralize icon updates
Impact: Improved maintainability and reduced code duplication
CSS Organization ✅
Problem: Universal selector () was in enhanced-styles.css instead of main CSS
Solution: Moved universal selector to main styles.css for better architecture
Impact: Better separation of concerns and improved CSS organization

TechnoBlogger14o3 committed:
Address code review feedback from Copilot AI and maintainer

Fixes based on PR #51 feedback:

Performance Improvements:
- Reduced progress image update frequency from 500ms to 5000ms (5 seconds)
- This significantly improves performance by reducing unnecessary API calls

Security Enhancements:
- Replaced innerHTML usage with safer DOM manipulation methods
- Created updateThemeIcon() helper function to safely update icon classes
- Eliminated potential XSS vulnerabilities from HTML injection

Code Quality Improvements:
- Extracted HTML injection logic into reusable helper function
- Reduced code duplication in theme icon updates
- Improved maintainability and readability

CSS Organization:
- Moved universal selector (*) from enhanced-styles.css to main styles.css
- Better separation of concerns between main styles and enhancements
- Improved CSS architecture and maintainability

Technical Details:
- updateThemeIcon() function safely manipulates DOM elements
- Uses querySelector and className updates instead of innerHTML
- Maintains fallback behavior for edge cases
- Preserves all existing functionality while improving security

All changes maintain backward compatibility and existing functionality
while addressing the specific concerns raised in the code review.

TechnoBlogger14o3 committed:
Address additional Copilot AI code review feedback

Additional fixes for year-in-progress PR #51 based on latest Copilot feedback:

Modern JavaScript Best Practices:
- Replaced 'var' declarations with 'let' and 'const' for better block scoping
- Updated queryString variable from 'var' to 'let' in theme.js
- Updated startDate and endDate variables from 'var' to 'let' in changeDate function

Memory Leak Prevention:
- Added proper cleanup for setInterval in theme.js (progressImgIntervalId)
- Added proper cleanup for setInterval in animation.js (cubeScaleIntervalId)
- Implemented beforeunload event listeners to clear intervals on page unload
- Prevents memory leaks in single-page applications and long-running sessions

Code Quality Improvements:
- Better variable scoping with modern JavaScript practices
- Proper resource cleanup to prevent memory accumulation
- Enhanced code maintainability and performance
- Follows modern JavaScript best practices and standards

Technical Details:
- Stored interval IDs in constants for proper cleanup
- Added beforeunload event listeners for both theme and animation intervals
- Maintains all existing functionality while improving resource management
- Prevents potential memory leaks in production environments

All changes maintain backward compatibility and existing functionality
while addressing the latest code review feedback for better code quality.
